packages/react-components/react-table/src/hooks/useColumnResizeState.ts

```diff
@@ -0,0 +1,85 @@
+import { useIsomorphicLayoutEffect } from '@fluentui/react-utilities';
+import * as React from 'react';
+import {
+  TableColumnDefinition,
+  TableColumnId,
+  ColumnResizeState,
+  ColumnWidthState,
+  UseColumnSizingParams,
+} from './types';
+import {
+  columnDefinitionsToState,
+  adjustColumnWidthsToFitContainer,
+  getColumnById,
+  setColumnProperty,
+  getTotalWidth,
+  getLastColumn,
+  getColumnByIndex,
+  getColumnWidth,
+  getLength,
+} from '../utils/columnResizeUtils';
+
+export function useColumnResizeState<T>(
+  columns: TableColumnDefinition<T>[],
+  containerWidth: number,
+  params: UseColumnSizingParams = {},
+): ColumnResizeState {
+  const { onColumnResize, columnSizingOptions } = params;
+
+  const [state, setState] = React.useState<ColumnWidthState[]>(
+    columnDefinitionsToState(columns, undefined, columnSizingOptions),
+  );
+
+  // Use layout effect here to make sure that updated columns receive proper styles immediately
+  useIsomorphicLayoutEffect(() => {
+    const intermediateState = columnDefinitionsToState(columns, state, columnSizingOptions);
+    setState(adjustColumnWidthsToFitContainer(intermediateState, containerWidth));
+  }, [columns, containerWidth, state, columnSizingOptions]);
+
+  const setColumnWidth = React.useCallback(
```

Comment:
```suggestion
  const setColumnWidth = useEventCallback(
```

Let's use and event callback instead so that the function reference does not change too often. Since it depends on state in the callback, it will more or less change reference every time anything in state changes.

This hook should generally be used in event handlers so it should never be called at render time.